File: drills/python/solutions/nc98_longest_increasing_path__by_value_order.py
def longestIncreasingPath(matrix):
    if not matrix or not matrix[0]:
        return 0

    # The same acyclicity used the other way round: process the squares from
    # largest value to smallest, and by the time a square is reached every
    # square it can move to has already been settled. A topological order
    # without ever building the graph -- sorting by value *is* the order.
    cells = [
        (matrix[r][c], r, c) for r in range(len(matrix)) for c in range(len(matrix[0]))
    ]
    lengths = {}

    for value, r, c in sorted(cells, reverse=True):
        best = 1
        for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
            nr, nc = r + dr, c + dc
            if 0 <= nr < len(matrix) and 0 <= nc < len(matrix[0]):
                if matrix[nr][nc] > value:
                    best = max(best, 1 + lengths[(nr, nc)])
        lengths[(r, c)] = best

    return max(lengths.values())

File: drills/go/solutions/nc98_longest_increasing_path__by_value_order.go
package main

import "sort"

func longestIncreasingPath(matrix [][]int) int {
	rows, cols := len(matrix), len(matrix[0])
	// Process cells in increasing value order: by the time a cell is
	// reached, every smaller neighbour already knows its best path, so no
	// recursion is needed.
	cells := make([][2]int, 0, rows*cols)
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			cells = append(cells, [2]int{r, c})
		}
	}
	sort.Slice(cells, func(i, j int) bool {
		return matrix[cells[i][0]][cells[i][1]] < matrix[cells[j][0]][cells[j][1]]
	})
	longest := make([][]int, rows)
	for r := range longest {
		longest[r] = make([]int, cols)
	}
	answer := 0
	for _, cell := range cells {
		r, c := cell[0], cell[1]
		best := 1
		for _, d := range [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}} {
			nr, nc := r+d[0], c+d[1]
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols && matrix[nr][nc] < matrix[r][c] {
				best = max(best, 1+longest[nr][nc])
			}
		}
		longest[r][c] = best
		answer = max(answer, best)
	}
	return answer
}
